File: built-in plugins dsp/mb-master-tilt.cpp
/**
 * MB Master Tilt EQ
 * Category : effect
 * Type     : mastering
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Single-knob tilt EQ for quick tonal balance adjustments
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MASTER_TILT_H
#define MB_MASTER_TILT_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMasterTilt : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-master-tilt";
    static constexpr const char* PLUGIN_NAME    = "MB Master Tilt EQ";
    static constexpr const char* PLUGIN_TYPE    = "mastering";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float tilt = 0f;  // range [-6, 6]
    float pivot = 1000f;  // range [200, 5000]
    float slope = 0.5f;  // range [0.1, 2]
    };

    MbMasterTilt() = default;
    ~MbMasterTilt() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.tilt = std::clamp(params.tilt, -6f, 6f);
        params.pivot = std::clamp(params.pivot, 200f, 5000f);
        params.slope = std::clamp(params.slope, 0.1f, 2f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Master Tilt EQ
        return input;
    }
};

#endif // MB_MASTER_TILT_H

File: built-in plugins dsp/mb-master-tilt.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbMasterTiltPlugin: PluginDefinition = { id: 'mb-master-tilt', slug: 'mb-master-tilt', name: 'MB Master Tilt EQ', category: 'effect', type: 'mastering' as any, version: '1.0.0', description: 'Single-knob tilt EQ for quick tonal balance adjustments', author: 'Max Booster', grade: 'A', parameters: [{ id: 'tilt', name: 'Tilt', type: 'float', defaultValue: 0, minValue: -6, maxValue: 6, automatable: true }, { id: 'pivot', name: 'Pivot Frequency', type: 'float', defaultValue: 1000, minValue: 200, maxValue: 5000, automatable: true }, { id: 'slope', name: 'Slope', type: 'float', defaultValue: 0.5, minValue: 0.1, maxValue: 2, automatable: true }], defaultPreset: { tilt: 0, pivot: 1000, slope: 0.5 } };

export default MbMasterTiltPlugin;
